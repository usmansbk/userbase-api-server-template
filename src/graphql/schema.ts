import { makeExecutableSchema } from "@graphql-tools/schema";

import authDirectiveTransformer from "./directives/auth";
import resolvers from "./resolvers";
import typeDefs from "./typeDefs";

const schema = authDirectiveTransformer(
  makeExecutableSchema({
    typeDefs,
    resolvers,
  }),
  "auth",
);

export default schema;
