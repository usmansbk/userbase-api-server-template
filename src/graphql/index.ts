import { ApolloServer } from "@apollo/server";
import { ApolloServerPluginDrainHttpServer } from "@apollo/server/plugin/drainHttpServer";
import type { Express } from "express";
import http from "http";
import type { AppContext } from "types";

import schema from "./schema";
import useWebSocketServer from "./ws";

export default async function createApolloHTTPServer(app: Express) {
  const httpServer = http.createServer(app);
  const serverCleanup = useWebSocketServer(schema, httpServer);

  const apolloServer = new ApolloServer<AppContext>({
    schema,
    plugins: [
      ApolloServerPluginDrainHttpServer({ httpServer }),
      {
        async serverWillStart() {
          return {
            async drainServer() {
              await serverCleanup.dispose();
            },
          };
        },
      },
    ],
    allowBatchedHttpRequests: true,
  });

  // Ensure we wait for our server to start
  await apolloServer.start();

  return { apolloServer, httpServer };
}
