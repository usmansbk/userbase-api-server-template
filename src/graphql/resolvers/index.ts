import path from "path";
import { mergeResolvers } from "@graphql-tools/merge";
import { loadFilesSync } from "@graphql-tools/load-files";

const resolversArray = loadFilesSync(path.join(__dirname, "."), {
  recursive: true,
  ignoreIndex: true,
});

export default mergeResolvers(resolversArray);
