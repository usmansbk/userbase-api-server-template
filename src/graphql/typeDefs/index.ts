import { loadFilesSync } from "@graphql-tools/load-files";
import { mergeTypeDefs } from "@graphql-tools/merge";
import path from "path";

const typesArray = loadFilesSync(path.join(__dirname, "."), {
  recursive: true,
  ignoreIndex: true,
  extensions: ["gql", "ts", "js"],
});

export default mergeTypeDefs(typesArray);
