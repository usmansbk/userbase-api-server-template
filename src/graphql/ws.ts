import { useServer } from "graphql-ws/lib/use/ws";
import { WebSocketServer } from "ws";
import i18next from "i18next";
import redisClient, { pubsub } from "@/config/redis";
import getPrismaClient from "@/config/database";
import smsClient from "@/utils/sms";
import logger from "@/utils/logger";
import storage from "@/utils/storage";
import type { IncomingMessage, ServerResponse, Server } from "http";
import type { GraphQLSchema } from "graphql";
import type { User } from "@prisma/client";
import type { AppContext } from "types";

export default function useWebSocketServer(
  schema: GraphQLSchema,
  server: Server<typeof IncomingMessage, typeof ServerResponse>,
) {
  const wsServer = new WebSocketServer({
    server,
    path: "/graphql",
  });

  return useServer(
    {
      schema,
      context: async (ctx): Promise<AppContext> => {
        let currentUser: User | undefined;

        const prismaClient = getPrismaClient();

        try {
          const token = ctx.connectionParams?.Authorization as
            | string
            | undefined;

          // TODO
          const decoded = token;
          if (decoded) {
            if (currentUser) {
              if (currentUser.language) {
                await i18next.changeLanguage(currentUser.language);
              }
            }
          }
        } catch (error) {
          logger.info({ error });
        }

        return {
          pubsub,
          smsClient,
          currentUser,
          redisClient,
          prismaClient,
          t: i18next.t,
          language: i18next.language,
          log: logger,
          storage,
        };
      },
      onConnect: (ctx) => {
        try {
          const token = ctx.connectionParams?.Authorization as
            | string
            | undefined;

          return !!token;
        } catch (error) {
          logger.info({ error });
        }
        return false;
      },
    },
    wsServer,
  );
}
