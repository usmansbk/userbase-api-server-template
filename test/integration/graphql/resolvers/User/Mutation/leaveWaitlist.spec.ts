import { createMockApolloServer } from "test/integration/graphql";
import createMockContext from "test/integration/graphql/context";

const query = `mutation LeaveWaitlist($email: EmailAddress!) {
			leaveWaitlist(email: $email) {
        success
				message
			}
		}`;

describe("Mutation.leaveWaitlist", () => {
  it("should remove user from the waitlist", async () => {
    const server = createMockApolloServer();
    const contextValue = await createMockContext();

    const userWaiting = await contextValue.prismaClient.user.create({
      data: {
        email: "test@email.com",
        firstName: "Testing",
        password: "test",
      },
    });

    const response = await server.executeOperation(
      {
        query,
        variables: {
          email: userWaiting.email,
        },
      },
      { contextValue },
    );

    const removed = await contextValue.prismaClient.user.findUnique({
      where: {
        id: userWaiting.id,
      },
    });

    expect(removed).toBeNull();
    expect(response).toMatchSnapshot();
  });

  it("should be idempotent", async () => {
    const server = createMockApolloServer();
    const contextValue = await createMockContext();

    const response = await server.executeOperation(
      {
        query,
        variables: {
          email: "test2@email.com",
        },
      },
      { contextValue },
    );

    expect(response).toMatchSnapshot();
  });
});
