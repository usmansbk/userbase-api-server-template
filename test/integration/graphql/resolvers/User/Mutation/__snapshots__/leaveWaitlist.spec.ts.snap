// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation.leaveWaitlist should be idempotent 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "leaveWaitlist": {
          "message": "You've left the waitlist.",
          "success": true,
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`Mutation.leaveWaitlist should remove user from the waitlist 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "leaveWaitlist": {
          "message": "You've left the waitlist.",
          "success": true,
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
